### ISSUE TITLE
DL4J and Keras predictions differ (Conv2D)

### ISSUE THREAD

**Issue by blackredscarf** [deeplearning4j/deeplearning4j#4756]
I have a network in keras, and I implement it in DL4j.
I try to use `layer.setParam` import the weights from keras for each layer,but the predictions is very different in same data. 

I check the weights and summary, it is no problem. I guess maybe some layers is error.

So I use `model.feedForward(data, false).get(layer)` to print the output of each layer, and compared with keras. I went looking for the first layer that made a huge difference and I found it.

I print its configuration, but I can not find the configuration error:

this layer configuration in keras:
```
{'activation': 'linear',
 'activity_regularizer': None,
 'bias_constraint': None,
 'bias_initializer': {'class_name': 'Zeros', 'config': {}},
 'bias_regularizer': None,
 'data_format': 'channels_last',
 'dilation_rate': (1, 1),
 'filters': 32,
 'kernel_constraint': None,
 'kernel_initializer': {'class_name': 'VarianceScaling',
  'config': {'distribution': 'uniform',
   'mode': 'fan_avg',
   'scale': 1.0,
   'seed': None}},
 'kernel_regularizer': None,
 'kernel_size': (3, 3),
 'name': 'Conv2d_1a_3x3',
 'padding': 'valid',
 'strides': (2, 2),
 'trainable': True,
 'use_bias': False}
```

this layer configuration in DL4j:
```
"Conv2d_2a_3x3" : {
  "LayerVertex" : {
	"layerConf" : {
	  "cacheMode" : "NONE",
	  "epochCount" : 0,
	  "iterationCount" : 0,
	  "l1ByParam" : {
		"W" : 0.0
	  },
	  "l2ByParam" : {
		"W" : 0.0
	  },
	  "layer" : {
		"convolution" : {
		  "activationFn" : {
			"Identity" : { }
		  },
		  "biasInit" : 0.0,
		  "biasUpdater" : null,
		  "constraints" : null,
		  "convolutionMode" : "Truncate",
		  "cudnnAlgoMode" : "PREFER_FASTEST",
		  "cudnnBwdDataAlgo" : null,
		  "cudnnBwdFilterAlgo" : null,
		  "cudnnFwdAlgo" : null,
		  "dilation" : [ 1, 1 ],
		  "dist" : null,
		  "gradientNormalization" : "None",
		  "gradientNormalizationThreshold" : 1.0,
		  "hasBias" : false,
		  "idropout" : {
			"@class" : "org.deeplearning4j.nn.conf.dropout.Dropout",
			"p" : 1.0,
			"pschedule" : null
		  },
		  "iupdater" : {
			"@class" : "org.nd4j.linalg.learning.config.Sgd",
			"learningRate" : 0.001
		  },
		  "kernelSize" : [ 3, 3 ],
		  "l1" : 0.0,
		  "l1Bias" : 0.0,
		  "l2" : 0.0,
		  "l2Bias" : 0.0,
		  "layerName" : "Conv2d_2a_3x3",
		  "nin" : 32,
		  "nout" : 32,
		  "padding" : [ 0, 0 ],
		  "stride" : [ 1, 1 ],
		  "weightInit" : "VAR_SCALING_NORMAL_FAN_AVG",
		  "weightNoise" : null
		}
	  },
	  "maxNumLineSearchIterations" : 5,
	  "miniBatch" : true,
	  "minimize" : true,
	  "optimizationAlgo" : "STOCHASTIC_GRADIENT_DESCENT",
	  "pretrain" : false,
	  "seed" : 123,
	  "stepFunction" : null,
	  "variables" : [ "W" ]
	},
	"outputVertex" : false,
	"preProcessor" : null
  }
},
```

Version: 0.9.1 / 0.9.2-SNAPSHOTS

This problem has troubled me for a long time, can you provide some suggestions?

**Comment by AlexDBlack** [comment]
Can you also post information about your hardware, and which backend (native, cuda, cuda+cudnn etc) you are using?

**Comment by blackredscarf** [comment]
@AlexDBlack 
Intel CPU
backend: native

**Comment by maxpumperla** [comment]
@blackredscarf I'm doing some general bug-hunting right now. I need more info to tackle this.

Can you provide me with a minimal example (ideally one conv layer with activations) for which you get discrepancies? If not, give me any example.

**Comment by maxpumperla** [comment]
@blackredscarf could you please provide me with your example?

**Comment by blackredscarf** [comment]
@maxpumperla 
the example: [example.rar](https://drive.google.com/file/d/1ERkOvYqtOA23_l5MsLHqsNev47MyNmh7/view?usp=sharing)

This is a complete example, maybe a little complicated.

#### python
python/exportWeights.ipynb	-	export the weights from the model

#### java
java/..../FaceInceptionResNetV1	-	the implement of model

#### scala
**scala/.../FaceNetTest	-	test enterance**
scala/.../LoadWeights	-	load the weights from file

#### resources
conf	-	the configuration of weights
weights -	the weights file

**Comment by maxpumperla** [comment]
@blackredscarf I tried to load your model back into keras, which fails due to bad marshalling. I assume you have some `Lambda` layers in there. Could you please tell me:

- what your model config looks like (json of your architecture)
- which layers feed into `Conv2d_2a_3x3`?

We have an e2e test on master for a conv net that checks predictions and gradients (and passes):
https://github.com/deeplearning4j/deeplearning4j/blob/9d782c260dfac22f8e6560cd66243080aba6e0d6/deeplearning4j-modelimport/src/test/java/org/deeplearning4j/nn/modelimport/keras/e2e/KerasModelEndToEndTest.java#L127

So I want to find out if this is due to some utility layer before that layer that produces wrong input to `Conv2d_2a_3x3`.

**Comment by blackredscarf** [comment]
@maxpumperla 
You must import this model in python3.6 and keras2.1+ environment. May report `UserWarning: No training configuration found in save`, but does not affect this model.

This is the summary of keras model:
https://gist.github.com/blackredscarf/8e64ba2663f6e6ec807883065ee4ef68

`Conv2d_2a_3x3` is the fifth layer. There is no Lambda layer in front of it.

**Comment by maxpumperla** [comment]
@blackredscarf ok, upon inspecting this a little more closely, few things:

- `Conv2d_2a_3x3` and `Conv2d_1a_3x3` are identical in their definition. if `Conv2d_2a_3x3` is off and `Conv2d_1a_3x3` isn't, I highly doubt that the problem comes from the conv layers.

- right before `Conv2d_2a_3x3` you have a batchnorm layer. I've seen that this layer has issues on import. For instance, no `momentum` or `scale` parameters are properly used. I need to fix that. I'm 98% your problem stems from this. To back my claim, note that we have a lot of prediction and grad checks for plain conv nets here:

https://github.com/deeplearning4j/deeplearning4j/blob/f50908f24c9f828973c4ccaf37283d7820d97d03/deeplearning4j-modelimport/src/test/java/org/deeplearning4j/nn/modelimport/keras/e2e/KerasModelEndToEndTest.java#L121-L142

- In your attempt to recreate the model in DL4J I also see a few things that don't match. For instance:

```
        String bn_name = _generate_layer_name("BatchNorm", name);
        graph.addLayer(bn_name, new BatchNormalization.Builder()
                        .activation(Activation.SIGMOID)
                        .eps(0.001).decay(0.995).l2(0)
                        .dropOut(new Dropout(1.0))
                        .updater(new Sgd(0.001))
                        .weightInit(WeightInit.XAVIER).build(), x);
```

why would you put Sigmoid activation here? It should be linear, since you follow up with a ReLU activation layer.

OK, depending on how this goes I might actually close this issue in favour of a more indicative batchnorm ticket.

**Comment by maxpumperla** [comment]
@blackredscarf I've created a test here
https://github.com/deeplearning4j/deeplearning4j/pull/5228/files#diff-54b41956f01cdb86d7a290961773a2df

that replicates the first 5 layers of your model (and tops it off with a `Flatten` and `Dense(1)` to predict single outputs). I predict random inputs in keras with fixed seed, import the model and input data into dl4j and predict there as well. there are no discrepancies.
